src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,652 @@
+where
+    P: Packet + Clone + Send + Sync + 'static,
+    F: Future<Output = Result<RespHeader, RpcError<String>>> + Unpin,
+{
+    /// Create a new receive header stream.
+    fn new(
+        client_inner: &'a RpcClientConnectionInner<P>,
+        interval: &'a mut Pin<Box<Interval>>,
+        recv_future: &'a mut F,
+    ) -> Self {
+        Self {
+            client_inner,
+            interval,
+            recv_future: Pin::new(recv_future),
+        }
+    }
+}
+
+impl<'a, P, F> Future for ReceiveHeaderFuture<'a, P, F>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+    F: Future<Output = Result<RespHeader, RpcError<String>>> + Unpin,
+{
+    type Output = Result<RespHeader, RpcError<String>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let self_mut: &mut ReceiveHeaderFuture<'a, P, F> = self.get_mut();
+        let client_inner = self_mut.client_inner;
+
+        // clean timeout tasks
+        let mut tick_ready = false;
+        while self_mut.interval.as_mut().poll_tick(cx).is_ready() {
+            // consume all ticks in once call
+            tick_ready = true;
+        }
+        if tick_ready {
+            let packets_keeper: &mut PacketsKeeper<P> = client_inner.get_packets_keeper_mut();
+            packets_keeper.clean_timeout_tasks();
+        }
+
+        // recv header data
+        self_mut.recv_future.as_mut().poll(cx)
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+#[allow(clippy::indexing_slicing)]
+mod tests {
+    use tokio::net::TcpListener;
+
+    use crate::storage::cache::rpc::packet::PacketStatus;
+
+    use super::*;
+    use std::time::Duration;
+
+    #[derive(Debug, Clone)]
+    pub struct TestPacket {
+        pub seq: u64,
+        pub op: u8,
+        pub status: PacketStatus,
+        pub timestamp: u64,
+    }
+
+    impl Packet for TestPacket {
+        fn seq(&self) -> u64 {
+            self.seq
+        }
+
+        fn set_seq(&mut self, seq: u64) {
+            self.seq = seq;
+        }
+
+        fn op(&self) -> u8 {
+            self.op
+        }
+
+        fn set_op(&mut self, op: u8) {
+            self.op = op;
+        }
+
+        fn set_timestamp(&mut self, timestamp: u64) {
+            self.timestamp = timestamp;
+        }
+
+        fn get_timestamp(&self) -> u64 {
+            self.timestamp
+        }
+
+        fn set_req_data(&mut self, _data: &[u8]) -> Result<(), RpcError<String>> {
+            Ok(())
+        }
+
+        fn get_req_data(&self) -> Result<Vec<u8>, RpcError<String>> {
+            Ok(Vec::new())
+        }
+
+        fn set_resp_data(&mut self, _data: &[u8]) -> Result<(), RpcError<String>> {
+            Ok(())
+        }
+
+        fn get_resp_data(&self) -> Result<Vec<u8>, RpcError<String>> {
+            Ok(Vec::new())
+        }
+
+        fn status(&self) -> PacketStatus {
+            self.status
+        }
+
+        fn set_status(&mut self, status: PacketStatus) {
```

Comment:
change to set_result(self,  Result<resp_data, Err()>)

Reply:
resp_data is in `set_resp_data`, changed to `        fn set_result(&self, _status: Result<(), RpcError>) {}`.